### PR TITLE
crimson,auth: fix FTBFS of crimson-osd and fix v1/v2 auth

### PR DESCRIPTION
**Pull request by tchaikov**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
cc @cyx1231st

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-04-27_02:20:42-rados-wip-kefu-testing-2019-04-26-2318-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-04-28_09:14:06-rados-wip-kefu-testing-2019-04-26-2318-distro-basic-smithi/

failed test tracked by https://tracker.ceph.com/issues/39449

File: src/crimson/mon/MonClient.cc [hunk 1]
```diff
@@ -525,6 +525,13 @@ int Client::handle_auth_request(ceph::net::ConnectionRef con,
                                 const ceph::bufferlist& payload,
                                 ceph::bufferlist *reply)
 {
+  // for some channels prior to nautilus (osd heartbeat), we tolerate the lack of
+  // an authorizer.
+  if (payload.length() == 0 &&
+      !con->get_messenger()->get_require_authorizer()) {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc]: updated to be in sync with #27832

File: src/crimson/net/ProtocolV1.cc [hunk 2]
```diff
@@ -249,13 +286,15 @@ ProtocolV1::repeat_connect()
...
-  if (h.authorizer) {
-    h.connect.authorizer_protocol = h.authorizer->protocol;
-    h.connect.authorizer_len = h.authorizer->bl.length();
+  bufferlist auth_bl = get_auth_payload();
+  if (auth_bl.length()) {
+    h.connect.authorizer_protocol = auth_meta->auth_method;
+    h.connect.authorizer_len = auth_bl.length();
     bl.append(create_static(h.connect));
-    bl.append(h.authorizer->bl);
+    bl.append(auth_bl.c_str(), auth_bl.length());
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/net/ProtocolV1.cc]: yeah, we should use `claim_append()`!

fixed.

File: src/crimson/net/ProtocolV1.cc [hunk 3]
```diff
@@ -494,26 +534,40 @@ seastar::future<stop_t> ProtocolV1::repeat_handle_connect()
...
-                                             h.connect.authorizer_protocol,
-                                             authorizer);
-    }).then([this] (ceph::net::msgr_tag_t tag, bufferlist&& authorizer_reply) {
-      memset(&h.reply, 0, sizeof(h.reply));
-      if (tag) {
-        return send_connect_reply(tag, std::move(authorizer_reply));
+
+      bufferlist authorizer_reply;
+      auth_meta->auth_method = h.connect.authorizer_protocol;
+      bool more = (bool)auth_meta->authorizer_challenge;
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/net/ProtocolV1.cc]: fixed.

File: src/crimson/net/ProtocolV1.cc [hunk 4]
```diff
@@ -236,6 +255,24 @@ ProtocolV1::handle_connect_reply(msgr_tag_t tag)
   }
 }
 
+ceph::bufferlist ProtocolV1::get_auth_payload()
+{
+  if (conn.peer_type == CEPH_ENTITY_TYPE_MON) {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/net/ProtocolV1.cc]: comment added.

**Comment by tchaikov** [comment]
@rzarzynski fixed and repushed.

File: src/crimson/net/ProtocolV1.cc [hunk 5]
```diff
@@ -774,6 +777,16 @@ seastar::future<> ProtocolV1::read_message()
       ::decode(m.footer, p);
       auto msg = ::decode_message(nullptr, 0, m.header, m.footer,
                                   m.front, m.middle, m.data, nullptr);
+      if (!msg) {
```
Comment (cyx1231st) [inline src/crimson/net/ProtocolV1.cc]: thanks!
FYI, also noticed this in https://github.com/ceph/ceph/pull/27430/commits/adf6a40efd82a67d5fccbebc8f3dd10696fb2fe8 #815
Seems this implementation doesn't abort the protocol in fault() correctly. Will fix later.

**Pull request merged**